cpp/bench.cxx

```diff
@@ -223,6 +221,182 @@ void vdot_f64c_blas(simsimd_f64_t const* a, simsimd_f64_t const* b, simsimd_size
+    /* 34 */ 3, /* 35 */ 4, /* 36 */ 4, /* 37 */ 5,
+    /* 38 */ 3, /* 39 */ 4, /* 3a */ 4, /* 3b */ 5,
+    /* 3c */ 4, /* 3d */ 5, /* 3e */ 5, /* 3f */ 6,
+    /* 40 */ 1, /* 41 */ 2, /* 42 */ 2, /* 43 */ 3,
+    /* 44 */ 2, /* 45 */ 3, /* 46 */ 3, /* 47 */ 4,
+    /* 48 */ 2, /* 49 */ 3, /* 4a */ 3, /* 4b */ 4,
+    /* 4c */ 3, /* 4d */ 4, /* 4e */ 4, /* 4f */ 5,
+    /* 50 */ 2, /* 51 */ 3, /* 52 */ 3, /* 53 */ 4,
+    /* 54 */ 3, /* 55 */ 4, /* 56 */ 4, /* 57 */ 5,
+    /* 58 */ 3, /* 59 */ 4, /* 5a */ 4, /* 5b */ 5,
+    /* 5c */ 4, /* 5d */ 5, /* 5e */ 5, /* 5f */ 6,
+    /* 60 */ 2, /* 61 */ 3, /* 62 */ 3, /* 63 */ 4,
+    /* 64 */ 3, /* 65 */ 4, /* 66 */ 4, /* 67 */ 5,
+    /* 68 */ 3, /* 69 */ 4, /* 6a */ 4, /* 6b */ 5,
+    /* 6c */ 4, /* 6d */ 5, /* 6e */ 5, /* 6f */ 6,
+    /* 70 */ 3, /* 71 */ 4, /* 72 */ 4, /* 73 */ 5,
+    /* 74 */ 4, /* 75 */ 5, /* 76 */ 5, /* 77 */ 6,
+    /* 78 */ 4, /* 79 */ 5, /* 7a */ 5, /* 7b */ 6,
+    /* 7c */ 5, /* 7d */ 6, /* 7e */ 6, /* 7f */ 7,
+    /* 80 */ 1, /* 81 */ 2, /* 82 */ 2, /* 83 */ 3,
+    /* 84 */ 2, /* 85 */ 3, /* 86 */ 3, /* 87 */ 4,
+    /* 88 */ 2, /* 89 */ 3, /* 8a */ 3, /* 8b */ 4,
+    /* 8c */ 3, /* 8d */ 4, /* 8e */ 4, /* 8f */ 5,
+    /* 90 */ 2, /* 91 */ 3, /* 92 */ 3, /* 93 */ 4,
+    /* 94 */ 3, /* 95 */ 4, /* 96 */ 4, /* 97 */ 5,
+    /* 98 */ 3, /* 99 */ 4, /* 9a */ 4, /* 9b */ 5,
+    /* 9c */ 4, /* 9d */ 5, /* 9e */ 5, /* 9f */ 6,
+    /* a0 */ 2, /* a1 */ 3, /* a2 */ 3, /* a3 */ 4,
+    /* a4 */ 3, /* a5 */ 4, /* a6 */ 4, /* a7 */ 5,
+    /* a8 */ 3, /* a9 */ 4, /* aa */ 4, /* ab */ 5,
+    /* ac */ 4, /* ad */ 5, /* ae */ 5, /* af */ 6,
+    /* b0 */ 3, /* b1 */ 4, /* b2 */ 4, /* b3 */ 5,
+    /* b4 */ 4, /* b5 */ 5, /* b6 */ 5, /* b7 */ 6,
+    /* b8 */ 4, /* b9 */ 5, /* ba */ 5, /* bb */ 6,
+    /* bc */ 5, /* bd */ 6, /* be */ 6, /* bf */ 7,
+    /* c0 */ 2, /* c1 */ 3, /* c2 */ 3, /* c3 */ 4,
+    /* c4 */ 3, /* c5 */ 4, /* c6 */ 4, /* c7 */ 5,
+    /* c8 */ 3, /* c9 */ 4, /* ca */ 4, /* cb */ 5,
+    /* cc */ 4, /* cd */ 5, /* ce */ 5, /* cf */ 6,
+    /* d0 */ 3, /* d1 */ 4, /* d2 */ 4, /* d3 */ 5,
+    /* d4 */ 4, /* d5 */ 5, /* d6 */ 5, /* d7 */ 6,
+    /* d8 */ 4, /* d9 */ 5, /* da */ 5, /* db */ 6,
+    /* dc */ 5, /* dd */ 6, /* de */ 6, /* df */ 7,
+    /* e0 */ 3, /* e1 */ 4, /* e2 */ 4, /* e3 */ 5,
+    /* e4 */ 4, /* e5 */ 5, /* e6 */ 5, /* e7 */ 6,
+    /* e8 */ 4, /* e9 */ 5, /* ea */ 5, /* eb */ 6,
+    /* ec */ 5, /* ed */ 6, /* ee */ 6, /* ef */ 7,
+    /* f0 */ 4, /* f1 */ 5, /* f2 */ 5, /* f3 */ 6,
+    /* f4 */ 5, /* f5 */ 6, /* f6 */ 6, /* f7 */ 7,
+    /* f8 */ 5, /* f9 */ 6, /* fa */ 6, /* fb */ 7,
+    /* fc */ 6, /* fd */ 7, /* fe */ 7, /* ff */ 8};
+
+uint64_t lower_qword(const __m128i v) { return _mm_cvtsi128_si64(v); }
+
+uint64_t higher_qword(const __m128i v) { return lower_qword(_mm_srli_si128(v, 8)); }
+
+uint64_t simd_sum_epu64(const __m128i v) { return lower_qword(v) + higher_qword(v); }
+
+uint64_t simd_sum_epu64(const __m256i v) {
```

Comment:
I think modern compilers might do this without asking in some cases, but using `inline` might encourage it (and could help with these small functions).

Reply:
The changes I've suggested so far are just low hanging fruit though. Have you used profiling tools to find which lines of code each approach is spending the most time in?

Reply:
Most time is spent in the main loop computing CSAs. Sadly, I can't access hardware performance counters on those machines.